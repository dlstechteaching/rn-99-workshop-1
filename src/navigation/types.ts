import { NativeStackScreenProps } from "@react-navigation/native-stack";
import { IOnboardingItem } from "../types/onboarding";
import { Navigation } from "./Navigation";

export type AppStackNavigatorParamList = {
    [Navigation.Home] : undefined,
    [Navigation.A] : undefined,
    [Navigation.B] : undefined,
    [Navigation.C] : undefined,
    [Navigation.Onboarding] : {
        stepIndex: number,
        data: IOnboardingItem[],
    },
};

export type HomeStackScreenProps = NativeStackScreenProps<
    AppStackNavigatorParamList,
    typeof Navigation.Home
>;

export type AScreenProps = NativeStackScreenProps<
    AppStackNavigatorParamList,
    typeof Navigation.A
>;

export type BScreenProps = NativeStackScreenProps<
    AppStackNavigatorParamList,
    typeof Navigation.B
>;

export type CScreenProps = NativeStackScreenProps<
    AppStackNavigatorParamList,
    typeof Navigation.C
>;

export type OnboardingScreenProps = NativeStackScreenProps<
    AppStackNavigatorParamList,
    typeof Navigation.Onboarding
>;