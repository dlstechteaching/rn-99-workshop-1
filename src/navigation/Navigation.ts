export enum Navigation {
	Home = 'app.home',
	A = 'app.ex.a',
	B = 'app.ex.b',
	C = 'app.ex.c',
	Onboarding = 'app.onboarding',
}
