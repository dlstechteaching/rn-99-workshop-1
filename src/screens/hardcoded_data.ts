import images from "../assets";
import { IOnboardingItem } from "../types/onboarding";

export const dataForBScreen = [
    {
        uid: 0,
        type: 'DESIGN',
        title: 'Create a new landing page Hero image',
        desc: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ultricies commodo erat, vel laoreet mi porta eget. Donec blandit orci mollis ex auctor, in hendrerit nunc pretium',
        badges: ['#Design', '#Visual', '#SPRINT 2'],
        comments: ['test', 'test', 'test']
    },
    {
        uid: 1,
        type: 'DESIGN',
        title: 'Create a new landing page Hero image',
        desc: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ultricies commodo erat, vel laoreet mi porta eget. Donec blandit orci mollis ex auctor, in hendrerit nunc pretium',
        badges: ['#Design', '#Visual', '#SPRINT 2'],
        comments: ['test', 'test', 'test']
    },
    {
        uid: 2,
        type: 'DESIGN',
        title: 'Create a new landing page Hero image',
        desc: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ultricies commodo erat, vel laoreet mi porta eget. Donec blandit orci mollis ex auctor, in hendrerit nunc pretium',
        badges: ['#Design', '#Visual', '#SPRINT 2'],
        comments: ['test', 'test', 'test']
    }
];

export const dataForCScreen = [
    {
      title: "General",
      data: [
          {
            iconName: 'ios-notifications-outline',
            title: 'Notifications',
            desc: 'Customize notifications'
          },
          {
            iconName: 'ellipsis-horizontal-outline',
            title: 'More customization',
            desc: 'Customize it more to fit your usage'
          },
        ]
    },
    {
      title: "Support",
      data: [
        {
          iconName: 'phone-portrait-outline',
          title: 'Contact',
        },
        {
          iconName: 'mail',
          title: 'Feedback',
        },
        {
            iconName: 'shield',
            title: 'Privacy Policy',
          },
          {
            iconName: 'alert-outline',
            title: 'About',
          },
      ]
    }
  ];

export const dataForOnboardingScreen: IOnboardingItem[]  = [
  {
    title: 'Welcome to monumental habits',
    desc: 'We can help you to be a better version of yourself',
    image: images.intro
  },
  {
    title: 'Create new habit easily',
    desc: 'We can help you to be a better version of yourself',
    image: images.habits
  },
  {
    title: 'Keep track of your progress',
    desc: 'We can help you to be a better version of yourself',
    image: images.progress
  },
  {
    title: 'Join a supportive community',
    desc: 'We can help you to be a better version of yourself',
    image: images.communitySupport
  }
];