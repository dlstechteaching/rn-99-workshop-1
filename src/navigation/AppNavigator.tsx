import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { HomeScreen } from '../screens/HomeScreen';
import { OnboardingScreen } from '../screens/OnboardingScreen';
import { ScreenA } from '../screens/ScreenA';
import { ScreenB } from '../screens/ScreenB';
import { ScreenC } from '../screens/ScreenC';
import { Navigation } from './Navigation';

const Stack = createNativeStackNavigator();

export const AppStack = () => {
	return (
		<Stack.Navigator initialRouteName={Navigation.Home} screenOptions={{ headerShown: false }}>
			<Stack.Screen name={Navigation.Home} component={HomeScreen} />
			<Stack.Screen name={Navigation.A} component={ScreenA} />
			<Stack.Screen name={Navigation.B} component={ScreenB} />
			<Stack.Screen name={Navigation.C} component={ScreenC} />
			<Stack.Screen name={Navigation.Onboarding} component={OnboardingScreen} />
		</Stack.Navigator>
	);
};
