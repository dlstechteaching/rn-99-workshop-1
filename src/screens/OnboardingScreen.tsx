import { TouchableOpacity, View, Text, Image, StyleSheet } from 'react-native';
import { StatusBar } from 'expo-status-bar';
import { MyStepper } from '../components/common/MyStepper';
import { useNavigation, useRoute } from '@react-navigation/native';
import { OnboardingScreenProps } from '../navigation/types';
import { Navigation } from '../navigation/Navigation';

export const OnboardingScreen: React.VFC = () => {
	const navigation = useNavigation<OnboardingScreenProps['navigation']>();
	const route = useRoute<OnboardingScreenProps['route']>();

	const data = route.params.data;
	const stepIndex = route.params.stepIndex;

	const currentStep = data[stepIndex];

	const navigateToStep = (step: number) =>
		navigation.navigate(Navigation.Onboarding, {
			stepIndex: step,
			data,
		});

	return (
		<View style={styles.container}>
			<StatusBar hidden />
			<Text style={styles.title} numberOfLines={2}>
				{currentStep.title.toUpperCase()}
			</Text>
			<Image source={currentStep.image as any} style={styles.image} />
			<Text style={styles.desc}>{currentStep.desc.toUpperCase()}</Text>
			<View style={styles.footer}>
				{stepIndex != data.length - 1 ? (
					<>
						<TouchableOpacity onPress={() => navigateToStep(data.length - 1)}>
							<Text>Skip</Text>
						</TouchableOpacity>
						<MyStepper stepIndex={stepIndex} nbOfSteps={data.length} width={'30%'} />
						<TouchableOpacity onPress={() => navigateToStep(stepIndex + 1)}>
							<Text>Next</Text>
						</TouchableOpacity>
					</>
				) : (
					<>
						<TouchableOpacity style={styles.cta} onPress={() => navigation.goBack()}>
							<Text>Get Started</Text>
						</TouchableOpacity>
					</>
				)}
			</View>
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1,
		flexDirection: 'column',
		padding: 40,
		width: '100%',
		backgroundColor: 'white',
		justifyContent: 'space-around',
		alignItems: 'center',
	},
	title: {
		marginTop: 20,
		fontSize: 22,
		textAlign: 'center',
		height: '7%',
	},
	image: {
		width: '80%',
		height: '40%',
		justifyContent: 'space-between',
		overflow: 'visible',
	},
	desc: {
		marginTop: 20,
		fontSize: 16,
		textAlign: 'center',
	},
	footer: {
		flexDirection: 'row',
		width: '100%',
		justifyContent: 'space-between',
	},
	cta: {
		padding: 20,
		width: '100%',
		alignItems: 'center',
		backgroundColor: 'rgb(241, 170, 102)',
		borderRadius: 10,
	},
});
