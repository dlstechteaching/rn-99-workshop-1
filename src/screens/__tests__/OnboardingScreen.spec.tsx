import { useNavigation, useRoute } from '@react-navigation/native';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { Navigation } from '../../navigation/Navigation';
import { dataForOnboardingScreen } from '../hardcoded_data';
import { OnboardingScreen } from '../OnboardingScreen';

jest.mock('@react-navigation/native');

describe('OnboardingScreen tests', () => {
	let component: ShallowWrapper;
	const navigationMock = {
		navigate: jest.fn(),
		goBack: jest.fn()
	};
	const routeMockDefault = {
		params: {
			stepIndex: 0,
			data: dataForOnboardingScreen
		}
	};

	beforeEach(() => {
		(useNavigation as jest.Mock).mockReturnValue(navigationMock);
	});

	it('should render properly', () => {
		givenUseRoute(routeMockDefault);
		givenComponent();
		thenItRendersProperly();
	});

	it('should render properly last step', () => {
		givenUseRoute({
			params: {
				stepIndex: dataForOnboardingScreen.length - 1,
				data: dataForOnboardingScreen
			}
		});
		givenComponent();
		thenItRendersProperly();
	});

	it('should skip and go to last step', () => {
		givenUseRoute(routeMockDefault);
		givenComponent();
		whenSkipButtonIsClicked();
		thenItNavigatesTo(3);
	});

	it('should go to next step', () => {
		givenUseRoute(routeMockDefault);
		givenComponent();
		whenNextButtonIsClicked();
		thenItNavigatesTo(1);
	});

	it('should go back in last step', () => {
		givenUseRoute({
			params: {
				stepIndex: dataForOnboardingScreen.length - 1,
				data: dataForOnboardingScreen
			}
		});
		givenComponent();
		whenCtaIsClicked();
		thenItGoesBack();
	});

	function givenComponent() {
		component = shallow(<OnboardingScreen />);
	}

	function givenUseRoute(routeMock: any) {
		(useRoute as jest.Mock).mockReturnValue(routeMock);
	}

	function whenSkipButtonIsClicked() {
		component.find('ForwardRef').at(0).simulate('press');
	}

	function whenNextButtonIsClicked() {
		component.find('ForwardRef').at(1).simulate('press');
	}

	function whenCtaIsClicked() {
		component.find('ForwardRef').at(0).simulate('press');
	}

	function thenItRendersProperly() {
		expect(component).toMatchSnapshot();
	}

	function thenItNavigatesTo(step: number){
		expect(navigationMock.navigate).toHaveBeenCalledWith(Navigation.Onboarding, {
			stepIndex: step,
			data: dataForOnboardingScreen
		});
	}

	function thenItGoesBack(){
		expect(navigationMock.goBack).toHaveBeenCalled();
	}

});