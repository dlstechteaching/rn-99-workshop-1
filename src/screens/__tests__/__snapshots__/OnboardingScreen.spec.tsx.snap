// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardingScreen tests should render properly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-around",
      "padding": 40,
      "width": "100%",
    }
  }
>
  <ExpoStatusBar
    hidden={true}
  />
  <Text
    numberOfLines={2}
    style={
      Object {
        "fontSize": 22,
        "height": "7%",
        "marginTop": 20,
        "textAlign": "center",
      }
    }
  >
    WELCOME TO MONUMENTAL HABITS
  </Text>
  <Image
    source={1}
    style={
      Object {
        "height": "40%",
        "justifyContent": "space-between",
        "overflow": "visible",
        "width": "80%",
      }
    }
  />
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginTop": 20,
        "textAlign": "center",
      }
    }
  >
    WE CAN HELP YOU TO BE A BETTER VERSION OF YOURSELF
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <ForwardRef
      onPress={[Function]}
    >
      <Text>
        Skip
      </Text>
    </ForwardRef>
    <MyStepper
      nbOfSteps={4}
      stepIndex={0}
      width="30%"
    />
    <ForwardRef
      onPress={[Function]}
    >
      <Text>
        Next
      </Text>
    </ForwardRef>
  </View>
</View>
`;

exports[`OnboardingScreen tests should render properly last step 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-around",
      "padding": 40,
      "width": "100%",
    }
  }
>
  <ExpoStatusBar
    hidden={true}
  />
  <Text
    numberOfLines={2}
    style={
      Object {
        "fontSize": 22,
        "height": "7%",
        "marginTop": 20,
        "textAlign": "center",
      }
    }
  >
    JOIN A SUPPORTIVE COMMUNITY
  </Text>
  <Image
    source={1}
    style={
      Object {
        "height": "40%",
        "justifyContent": "space-between",
        "overflow": "visible",
        "width": "80%",
      }
    }
  />
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginTop": 20,
        "textAlign": "center",
      }
    }
  >
    WE CAN HELP YOU TO BE A BETTER VERSION OF YOURSELF
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <ForwardRef
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(241, 170, 102)",
          "borderRadius": 10,
          "padding": 20,
          "width": "100%",
        }
      }
    >
      <Text>
        Get Started
      </Text>
    </ForwardRef>
  </View>
</View>
`;
